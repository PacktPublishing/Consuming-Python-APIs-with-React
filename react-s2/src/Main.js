import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Comp1 from './Comp1.js';
import Comp2 from './Comp2.js';

class Main extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.initialValue, title: "", explanation: "",
    }
  }
  async componentDidMount() {
    fetch("https://api.nasa.gov/planetary/apod?api_key=DEMO_KEY")
      .then(response => response.json())
      .then(
        data =>
        this.setState({title: data.title, explanation: data.explanation})
      )
      .catch((error) => {console.log('Error fetching and parsing data', error);});
  }
  inc_value() {
    const newValue = this.state.value + 1;
    this.setState({value: newValue,});
  }
  render() {
    return (
      <div className="App flex-container">
        <Comp1 text="Click me" inc_value={this.inc_value.bind(this)}/>
        <Comp2
          value={this.state.value} title={this.state.title}
          explanation={this.state.explanation}
        />
      </div>
    );
  }
  static propTypes = {
    initialValue: PropTypes.number.isRequired,
  }
}

export default Main;
