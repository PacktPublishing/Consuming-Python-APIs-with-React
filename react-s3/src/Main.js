import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Box, Grid, Grommet } from 'grommet';
import { grommetTheme } from './GrommetStyle.js';
import { BookmarksList } from './BookmarkList.js';
import { BookmarkDetail } from './BookmarkDetail.js';
import { Editor, editorMasks } from './Editor.js';

export default class Main extends Component {
  render () {
    let editorExistingItemData =
      ((this.props.editorState &
      (editorMasks.AMENDING | editorMasks.DELETING)) === 0) ? {} :
      this.props.bookmarkData;
    return (
      <Grommet theme={grommetTheme}>
        <Grid
          areas={[
            { name: 'main', start: [0, 0], end: [0, 0] },
            { name: 'detail', start: [1, 0], end: [1, 0] },
            { name: 'foot', start: [0, 1], end: [1, 1] },
          ]}
          columns={['medium', 'flex']}
          rows={['medium', 'small']}
          gap='small'
        >
          <Box
            gridArea='main' background='color-primary-0'
            overflow="scroll" pad="xxsmall"
          >
            <BookmarksList
              addBookmarkStart={this.props.addBookmarkStart}
              clientToken={this.props.clientToken}
              deleteBookmarkStart={this.props.deleteBookmarkStart}
              editBookmarkStart={this.props.editBookmarkStart}
              bookmarks={this.props.data}
              location={this.props.location} history={this.props.history}
              refreshThisBookmark={this.props.selectClick}
              selectClick={this.props.selectClick}
              selectedBookmark={this.props.bookmarkData}
            />
          </Box>
          <Box
            gridArea='detail' background='color-secondary-2-4'
            overflow="scroll" pad="xsmall"
          >
            <BookmarkDetail
              clientToken={this.props.clientToken}
              detail={this.props.bookmarkData}
              addNoteStart={this.props.addNoteStart}
              deleteNoteStart={this.props.deleteNoteStart}
              editNoteStart={this.props.editNoteStart}
              addCommentStart={this.props.addCommentStart}
              deleteCommentStart={this.props.deleteCommentStart}
              editCommentStart={this.props.editCommentStart}
              location={this.props.location} history={this.props.history}
            />
          </Box>
          <Box
            gridArea='foot' background='color-secondary-1-4'
            overflow="scroll" pad="xxsmall"
          >
            <Editor
              existingItemData={editorExistingItemData}
              state={this.props.editorState}
              stopEdit={this.props.stopEdit}
              clientToken={this.props.clientToken}
              itemId={this.props.alterId}
              refreshBookmarks={this.props.getBookmarks}
              refreshThisBookmark={this.props.selectClick}
              location={this.props.location} history={this.props.history}
              selectedBookmark={this.props.bookmarkData}
            />
          </Box>
        </Grid>
      </Grommet>
    )
  }
  static defaultProps = {
    alterId: null,
    selectedBookmarkId: null,
  }
  static propTypes = {
    addBookmarkStart: PropTypes.func.isRequired,
    addCommentStart: PropTypes.func.isRequired,
    addNoteStart: PropTypes.func.isRequired,
    alterId: PropTypes.number,
    bookmarkData: PropTypes.object,
    clientToken: PropTypes.object.isRequired,
    data: PropTypes.arrayOf(PropTypes.object),
    deleteBookmarkStart: PropTypes.func.isRequired,
    deleteCommentStart: PropTypes.func.isRequired,
    deleteNoteStart: PropTypes.func.isRequired,
    editBookmarkStart: PropTypes.func.isRequired,
    editCommentStart: PropTypes.func.isRequired,
    editNoteStart: PropTypes.func.isRequired,
    editorState: PropTypes.number.isRequired,
    getBookmarks: PropTypes.func.isRequired,
    selectClick: PropTypes.func.isRequired,
    selectedBookmarkId: PropTypes.number,
    stopEdit: PropTypes.func.isRequired,
  }
}
