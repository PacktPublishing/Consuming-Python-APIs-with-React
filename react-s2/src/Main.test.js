import React from 'react';
import ReactDOM from 'react-dom';
import Main from './Main';
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { MemoryRouter } from 'react-router-dom';
import { fetchMock } from 'fetch-mock';

// http://airbnb.io/enzyme/docs/installation/index.html
configure({ adapter: new Adapter() })

const sleep = (ms) => {
  return new Promise(resolve => setTimeout(resolve, ms));
}

test("inits state correctly and renders", () => {
  const wrapper = shallow(
    <Main initialValue={10} />
  )
  expect(wrapper.state()).toEqual({value: 10, title: "", explanation: ""});
  expect(wrapper).toMatchSnapshot();
});

describe('testing api', () => {

  test('obtains remote data (simulated)', async () => {
    fetchMock.get('*', { title: "ABC", explanation: "123" });
    const wrapper = mount(
      <MemoryRouter>
        <Main initialValue={8} />
      </MemoryRouter>
    )
    const mainWrapper = wrapper.find('Main')
    expect(mainWrapper.state()).toEqual(
      {value: 8, title: "", explanation: ""}
    );
    // Need a delay for componentDidMount to execute and subsequent render
    await sleep(200);

    expect(mainWrapper.state()).toEqual(
      {value: 8,  title: "ABC", explanation: "123"}
    );
    const Comp2Wrapper = mainWrapper.find('Comp2');
    expect(Comp2Wrapper.instance().props).toEqual(
      {value: 8,  title: "ABC", explanation: "123"}
    );
    fetchMock.restore();
  })

  test('correct API call', async () => {
    const wrapper = shallow(
      <Main initialValue={8} />
    )
    fetchMock.get('*', { title: "ABC", explanation: "123" });
    await wrapper.instance().componentDidMount();
    expect(fetchMock.lastCall()[0]).toEqual(
      "https://api.nasa.gov/planetary/apod?api_key=DEMO_KEY"
    )
    fetchMock.restore();
  })
})
