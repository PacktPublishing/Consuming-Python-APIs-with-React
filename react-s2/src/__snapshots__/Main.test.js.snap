// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inits state correctly and renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Main
    initialValue={10}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Comp1
          inc_value={[Function]}
          text="Click me"
        />,
        <Comp2
          explanation=""
          title=""
          value={10}
        />,
      ],
      "className": "App flex-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "inc_value": [Function],
          "text": "Click me",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "explanation": "",
          "title": "",
          "value": 10,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Comp1
            inc_value={[Function]}
            text="Click me"
          />,
          <Comp2
            explanation=""
            title=""
            value={10}
          />,
        ],
        "className": "App flex-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "inc_value": [Function],
            "text": "Click me",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "explanation": "",
            "title": "",
            "value": 10,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
